docs/proposals/garbage-collection.md

```diff
@@ -0,0 +1,382 @@
+* At last, removes the itself from the `ObjectMeta.Finalizers` of the object.
+
+# Related issues
+
+## Orphan adoption
+
+Controllers are responsible for adopting orphaned dependent resources. To do so, controllers
+* Checks a potential dependent object’s OwnerReferences to determine if it is orphaned.
+* Fills the OwnerReferences if the object matches the controller’s selector and is orphaned.
+
+There is a potential race between the "orphan" finalizer removing an owner reference and the controllers adding it back during adoption. Imagining this case: a user deletes an owning object and intends to orphan the dependent objects, so the GC removes the owner from the dependent object's OwnerReferences list, but the controller of the owner resource hasn't observed the deletion yet, so it adopts the dependent again and adds the reference back, resulting in the mistaken deletion of the dependent object. This race can be avoided by implementing Status.ObservedGeneration in all resources. Before updating the dependent Object's OwnerReferences, the "orphan" finalizer checks Status.ObservedGeneration of the owning object to ensure its controller has already observed the deletion.
+
+## Upgrading a cluster to support cascading deletion
+
+For the master, after upgrading to a version that supports cascading deletion, the OwnerReferences of existing objects remain empty, so the controllers will regard them as orphaned and start the adoption procedures. After the adoptions are done, server-side cascading will be effective for these existing objects.
+
+For nodes, cascading deletion does not affect them.
+
+For kubectl, we will keep the kubectl’s cascading deletion logic for one more release.
+
+# End-to-End Examples
+
+This section presents two examples of all components working together to enforce the cascading deletion or orphaning.
+
+## Life of a Deployment and its descendants
+
+1. User creates a deployment `D1`.
+2. The Propagator of the GC observes the creation. It creates an entry of `D1` in the DAG.
+3. The deployment controller observes the creation of `D1`. It creates the replicaset `R1`, whose OwnerReferences field contains a reference to `D1`, and has the "orphan" finalizer in its ObjectMeta.Finalizers map.
+4. The Propagator of the GC observes the creation of `R1`. It creates an entry of `R1` in the DAG, with `D1` as its owner.
+5. The replicaset controller observes the creation of `R1` and creates Pods `P1`~`Pn`, all with `R1` in their OwnerReferences.
+6. The Propagator of the GC observes the creation of `P1`~`Pn`. It creates entries for them in the DAG, with `R1` as their owner.
+
+  ***In case the user wants to cascadingly delete `D1`'s descendants, then***
+
+7. The user deletes the deployment `D1`, with `DeleteOptions.OrphanDependents=false`. API server checks if `D1` has "orphan" finalizer in its Finalizers map, if so, it updates `D1` to remove the "orphan" finalizer. Then API server deletes `D1`.
```

Comment:
If this happens, then the GC, upon a crash, has no way to observe the desired graceful deletion period of D1.  This may be acceptable in some cases (those members should have well defined grace periods of their own).  But it means that setting grace period on those items isn't guaranteed to propagate.

Setting OrphanDependents=false and GracePeriod=0 should apply graceperiod=0 to all descendents in most cases.